File: backend/app/lib/topic_utils.py
"""Topic hydration – map AI-quoted text to precise transcript segment boundaries.

Ported from vendors/longcut/lib/topic-utils.ts.
"""

from __future__ import annotations

import math
from typing import Any

from app.lib.quote_matcher import (
    TranscriptIndex,
    build_transcript_index,
    find_text_in_transcript,
)
from app.lib.timestamp_utils import parse_timestamp_range


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _coerce_number(value: Any) -> float | None:
    if isinstance(value, (int, float)) and math.isfinite(value):
        return float(value)
    if isinstance(value, str) and value.strip():
        try:
            v = float(value)
            return v if math.isfinite(v) else None
        except ValueError:
            return None
    return None


def _normalize_transcript(segments: list[dict] | None) -> list[dict]:
    if not isinstance(segments, list):
        return []
    result: list[dict] = []
    for seg in segments:
        if not seg:
            continue
        start = _coerce_number(seg.get("start"))
        dur = _coerce_number(seg.get("duration"))
        if start is None or dur is None:
            continue
        text = seg.get("text", "")
        if not isinstance(text, str):
            text = str(text) if text is not None else ""
        result.append({"text": text, "start": start, "duration": dur})
    return result


def _compute_duration(segments: list[dict]) -> float:
    total = 0.0
    for seg in segments:
        s = _coerce_number(seg.get("start"))
        e = _coerce_number(seg.get("end"))
        if s is not None and e is not None:
            total += max(0.0, e - s)
    return total


def _approximate_time_offset(segment: dict, char_offset: int | None) -> float:
    dur = _coerce_number(segment.get("duration"))
    if dur is None or dur <= 0:
        return 0.0
    text = segment.get("text", "")
    if not isinstance(char_offset, (int, float)) or not text:
        return 0.0
    safe_len = max(1, len(text))
    clamped = max(0, min(int(char_offset), safe_len))
    return dur * (clamped / safe_len)


def _find_segment_index_by_time(transcript: list[dict], time: float) -> int:
    if not math.isfinite(time):
        return -1
    for i, seg in enumerate(transcript):
        seg_end = seg["start"] + seg["duration"]
        if seg["start"] <= time <= seg_end:
            return i
        if time < seg["start"]:
            return max(0, i - 1)
    return len(transcript) - 1 if transcript else -1


def _create_segment_from_match(
    match: Any, transcript: list[dict], preferred_text: str | None = None
) -> dict | None:
    start_seg = transcript[match.start_segment_idx] if match.start_segment_idx < len(transcript) else None
    end_seg = transcript[match.end_segment_idx] if match.end_segment_idx < len(transcript) else None
    if not start_seg or not end_seg:
        return None

    start_offset = _approximate_time_offset(start_seg, match.start_char_offset)
    end_char = match.end_char_offset if isinstance(match.end_char_offset, (int, float)) else len(end_seg["text"])
    end_offset = _approximate_time_offset(end_seg, end_char)

    start_time = start_seg["start"] + start_offset
    end_time = end_seg["start"] + end_offset
    if not math.isfinite(end_time) or end_time <= start_time:
        end_time = end_seg["start"] + end_seg.get("duration", 0)
        if end_time <= start_time:
            end_time = start_time + max(5, end_seg.get("duration", 0))

    fallback_text = " ".join(
        s["text"] for s in transcript[match.start_segment_idx : match.end_segment_idx + 1]
    ).strip()

    return {
        "start": start_time,
        "end": end_time,
        "text": (preferred_text or fallback_text or "").strip(),
        "startSegmentIdx": match.start_segment_idx,
        "endSegmentIdx": match.end_segment_idx,
        "startCharOffset": match.start_char_offset if isinstance(match.start_char_offset, int) else 0,
        "endCharOffset": int(end_char),
        "hasCompleteSentences": match.match_strategy != "fuzzy-ngram",
    }


def _create_segment_from_timestamp(
    timestamp: str | None, transcript: list[dict], preferred_text: str | None = None
) -> dict | None:
    rng = parse_timestamp_range(timestamp) if timestamp else None
    if not rng:
        return None
    range_start, range_end = rng

    start_idx = _find_segment_index_by_time(transcript, range_start)
    end_idx = _find_segment_index_by_time(transcript, range_end)
    if start_idx == -1:
        start_idx = 0
    if end_idx == -1:
        end_idx = len(transcript) - 1
    if end_idx < start_idx:
        end_idx = start_idx

    start_seg = transcript[start_idx]
    end_seg = transcript[end_idx]

    start_time = max(start_seg["start"], min(range_start, start_seg["start"] + start_seg["duration"]))
    end_time = min(
        end_seg["start"] + end_seg["duration"],
        max(range_end, start_time + max(5, end_seg.get("duration", 0))),
    )
    if end_time <= start_time:
        end_time = start_time + max(5, end_seg.get("duration", 0))

    combined_text = " ".join(s["text"] for s in transcript[start_idx : end_idx + 1]).strip()

    return {
        "start": start_time,
        "end": end_time,
        "text": (preferred_text or combined_text or "").strip(),
        "startSegmentIdx": start_idx,
        "endSegmentIdx": end_idx,
        "startCharOffset": 0,
        "endCharOffset": len(end_seg["text"]),
        "hasCompleteSentences": False,
    }


def _create_fallback_segment(transcript: list[dict]) -> dict | None:
    if not transcript:
        return None
    start = transcript[0]
    end_idx = 0
    end_time = start["start"]
    for i, seg in enumerate(transcript):
        end_idx = i
        end_time = seg["start"] + seg["duration"]
        if end_time - start["start"] >= 60:
            break
    combined = " ".join(s["text"] for s in transcript[: end_idx + 1]).strip()
    return {
        "start": start["start"],
        "end": end_time,
        "text": combined,
        "startSegmentIdx": 0,
        "endSegmentIdx": end_idx,
        "startCharOffset": 0,
        "endCharOffset": len(transcript[end_idx]["text"]),
        "hasCompleteSentences": False,
    }


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def hydrate_topics_with_transcript(
    topics: list[dict] | None,
    transcript: list[dict] | None,
    index: TranscriptIndex | None = None,
) -> list[dict]:
    """Map AI-generated topics to precise transcript segment boundaries.

    Each topic dict should contain at minimum ``title`` and optionally
    ``quote: {"timestamp": "[MM:SS-MM:SS]", "text": "..."}``.

    Returns topics enriched with ``segments`` and ``duration``.
    """
    if not isinstance(topics, list) or not topics:
        return topics if isinstance(topics, list) else []

    norm_transcript = _normalize_transcript(transcript)
    has_transcript = len(norm_transcript) > 0

    if has_transcript and index is None:
        index = build_transcript_index(norm_transcript)

    result: list[dict] = []
    for topic in topics:
        segments: list[dict] = []

        if has_transcript and index is not None:
            quote = topic.get("quote")
            quote_text = quote.get("text", "") if isinstance(quote, dict) else ""
            quote_ts = quote.get("timestamp") if isinstance(quote, dict) else None

            # Strategy 1: match quote text
            if quote_text:
                match = find_text_in_transcript(
                    norm_transcript,
                    quote_text,
                    index,
                    min_similarity=0.75,
                )
                if match:
                    seg = _create_segment_from_match(match, norm_transcript, quote_text)
                    if seg:
                        segments = [seg]

            # Strategy 2: fallback to timestamp range
            if not segments:
                seg = _create_segment_from_timestamp(quote_ts, norm_transcript, quote_text)
                if seg:
                    segments = [seg]

            # Strategy 3: first 60s fallback
            if not segments:
                seg = _create_fallback_segment(norm_transcript)
                if seg:
                    segments = [seg]

        duration_secs = _compute_duration(segments)
        duration = round(duration_secs) if duration_secs > 0 else topic.get("duration", 0)

        result.append({**topic, "segments": segments, "duration": duration})

    return result

File: lib/topic-utils.ts
import type { Topic, TopicCandidate, TranscriptSegment } from '@/lib/types';
import {
  buildTranscriptIndex,
  findTextInTranscript,
  type TranscriptIndex,
} from '@/lib/quote-matcher';
import { parseTimestampRange } from '@/lib/timestamp-utils';

type TopicSegment = Topic['segments'][number];

export function topicTotalDuration(topic: Topic): number {
  return topic.segments.reduce((acc, s) => acc + (s.end - s.start), 0);
}

export function normalizeTopicQuoteText(text: string): string {
  return text.replace(/\s+/g, ' ').trim().toLowerCase();
}

export function topicQuoteKey(topic: {
  quote?: { timestamp?: string; text?: string };
}): string | null {
  if (!topic.quote?.timestamp || !topic.quote.text) return null;
  const normalizedText = normalizeTopicQuoteText(topic.quote.text);
  if (!normalizedText) return null;
  return `${topic.quote.timestamp}|${normalizedText}`;
}

function coerceNumber(value: unknown): number | null {
  if (typeof value === 'number' && Number.isFinite(value)) return value;
  if (typeof value === 'string' && value.trim()) {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) return parsed;
  }
  return null;
}

function normalizeSegment(segment: TopicSegment | null | undefined): TopicSegment | null {
  if (!segment) return null;
  const start = coerceNumber((segment as { start?: unknown }).start);
  const end = coerceNumber((segment as { end?: unknown }).end);
  if (start === null || end === null || end < start) return null;

  const startSegmentIdx = coerceNumber((segment as { startSegmentIdx?: unknown }).startSegmentIdx);
  const endSegmentIdx = coerceNumber((segment as { endSegmentIdx?: unknown }).endSegmentIdx);
  const startCharOffset = coerceNumber((segment as { startCharOffset?: unknown }).startCharOffset);
  const endCharOffset = coerceNumber((segment as { endCharOffset?: unknown }).endCharOffset);

  return {
    ...segment,
    start,
    end,
    startSegmentIdx: startSegmentIdx ?? undefined,
    endSegmentIdx: endSegmentIdx ?? undefined,
    startCharOffset: startCharOffset ?? undefined,
    endCharOffset: endCharOffset ?? undefined,
  };
}

function normalizeSegments(segments: TopicSegment[] | null | undefined): TopicSegment[] {
  if (!Array.isArray(segments)) return [];
  return segments.map(normalizeSegment).filter((segment): segment is TopicSegment => !!segment);
}

function normalizeTranscriptSegment(segment: TranscriptSegment | null | undefined): TranscriptSegment | null {
  if (!segment) return null;
  const start = coerceNumber((segment as { start?: unknown }).start);
  const duration = coerceNumber((segment as { duration?: unknown }).duration);
  if (start === null || duration === null) return null;

  return {
    text: typeof segment.text === 'string' ? segment.text : String(segment.text ?? ''),
    start,
    duration,
    translatedText: segment.translatedText,
  };
}

export function normalizeTranscript(
  transcript: TranscriptSegment[] | null | undefined,
): TranscriptSegment[] {
  if (!Array.isArray(transcript)) return [];
  return transcript
    .map(normalizeTranscriptSegment)
    .filter((segment): segment is TranscriptSegment => !!segment);
}

function approximateTimeOffset(segment: TranscriptSegment | undefined, charOffset: number): number {
  if (!segment?.text || !Number.isFinite(segment.duration) || segment.duration <= 0) return 0;
  const ratio = Math.max(0, Math.min(charOffset, segment.text.length)) / Math.max(1, segment.text.length);
  return segment.duration * ratio;
}

function segmentFromMatch(
  match: NonNullable<ReturnType<typeof findTextInTranscript>>,
  transcript: TranscriptSegment[],
  preferredText?: string,
): TopicSegment | null {
  const startSegment = transcript[match.startSegmentIdx];
  const endSegment = transcript[match.endSegmentIdx];
  if (!startSegment || !endSegment) return null;

  const start = startSegment.start + approximateTimeOffset(startSegment, match.startCharOffset);
  let end = endSegment.start + approximateTimeOffset(endSegment, match.endCharOffset);
  if (!Number.isFinite(end) || end <= start) {
    end = endSegment.start + endSegment.duration;
  }
  if (end <= start) {
    end = start + Math.max(5, endSegment.duration || 0);
  }

  const text = transcript
    .slice(match.startSegmentIdx, match.endSegmentIdx + 1)
    .map((segment) => segment.text)
    .join(' ')
    .trim();

  return {
    start,
    end,
    text: (preferredText || text || '').trim(),
    startSegmentIdx: match.startSegmentIdx,
    endSegmentIdx: match.endSegmentIdx,
    startCharOffset: match.startCharOffset,
    endCharOffset: match.endCharOffset,
    hasCompleteSentences: !match.matchStrategy.startsWith('fuzzy'),
    confidence: match.confidence,
  };
}

function findSegmentIndexByTime(transcript: TranscriptSegment[], time: number): number {
  if (!Number.isFinite(time)) return -1;
  for (let i = 0; i < transcript.length; i++) {
    const segment = transcript[i];
    const end = segment.start + segment.duration;
    if (time >= segment.start && (time < end || i === transcript.length - 1)) return i;
    if (time < segment.start) return Math.max(0, i - 1);
  }
  return transcript.length - 1;
}

function segmentFromTimestamp(
  timestamp: string | undefined,
  transcript: TranscriptSegment[],
  preferredText?: string,
): TopicSegment | null {
  if (!timestamp || !transcript.length) return null;
  const range = parseTimestampRange(timestamp);
  if (!range) return null;

  let startIdx = findSegmentIndexByTime(transcript, range.start);
  let endIdx = findSegmentIndexByTime(transcript, range.end);
  if (startIdx < 0) startIdx = 0;
  if (endIdx < 0) endIdx = transcript.length - 1;
  if (endIdx < startIdx) endIdx = startIdx;

  const startSegment = transcript[startIdx];
  const endSegment = transcript[endIdx];
  if (!startSegment || !endSegment) return null;

  const combinedText = transcript
    .slice(startIdx, endIdx + 1)
    .map((segment) => segment.text)
    .join(' ')
    .trim();
  const start = Math.max(startSegment.start, Math.min(range.start, startSegment.start + startSegment.duration));
  let end = Math.min(endSegment.start + endSegment.duration, Math.max(range.end, start));
  if (end <= start) end = start + Math.max(5, endSegment.duration || 0);

  return {
    start,
    end,
    text: (preferredText || combinedText || '').trim(),
    startSegmentIdx: startIdx,
    endSegmentIdx: endIdx,
    startCharOffset: 0,
    endCharOffset: endSegment.text.length,
    hasCompleteSentences: false,
    confidence: 0.5,
  };
}

function fallbackSegment(transcript: TranscriptSegment[]): TopicSegment | null {
  if (!transcript.length) return null;
  const startSegment = transcript[0];
  let endIdx = 0;
  let end = startSegment.start + startSegment.duration;

  for (let i = 0; i < transcript.length; i++) {
    endIdx = i;
    end = transcript[i].start + transcript[i].duration;
    if (end - startSegment.start >= 60) break;
  }

  return {
    start: startSegment.start,
    end,
    text: transcript
      .slice(0, endIdx + 1)
      .map((segment) => segment.text)
      .join(' ')
      .trim(),
    startSegmentIdx: 0,
    endSegmentIdx: endIdx,
    startCharOffset: 0,
    endCharOffset: transcript[endIdx].text.length,
    hasCompleteSentences: false,
    confidence: 0,
  };
}

function getTranscriptIndex(existing: TranscriptIndex | null, transcript: TranscriptSegment[]): TranscriptIndex {
  return existing ?? buildTranscriptIndex(transcript);
}

function computeDuration(segments: TopicSegment[]): number {
  return Math.round(
    segments.reduce((total, segment) => total + Math.max(0, segment.end - segment.start), 0),
  );
}

export function hydrateTopicsWithTranscript(
  topics: Topic[] | null | undefined,
  transcript: TranscriptSegment[] | null | undefined,
): Topic[] {
  if (!Array.isArray(topics)) return [];
  if (!topics.length) return topics;

  const normalizedTranscript = normalizeTranscript(transcript);
  if (!normalizedTranscript.length) {
    return topics.map((topic) => ({
      ...topic,
      segments: normalizeSegments(topic.segments),
    }));
  }

  let transcriptIndex: TranscriptIndex | null = null;

  return topics.map((topic) => {
    let hydratedSegments = normalizeSegments(topic.segments);

    if (!hydratedSegments.length) {
      const quoteText = topic.quote?.text?.trim();
      if (quoteText) {
        transcriptIndex = getTranscriptIndex(transcriptIndex, normalizedTranscript);
        const match = findTextInTranscript(normalizedTranscript, quoteText, transcriptIndex, {
          strategy: 'all',
          minSimilarity: 0.8,
          maxSegmentWindow: 20,
        });

        if (match) {
          const segment = segmentFromMatch(match, normalizedTranscript, quoteText);
          if (segment) hydratedSegments = [segment];
        }

        if (!hydratedSegments.length) {
          const range = parseTimestampRange(topic.quote?.timestamp ?? '');
          if (range) {
            const startIdx = Math.max(0, findSegmentIndexByTime(normalizedTranscript, range.start) - 2);
            const endIdx = Math.max(startIdx, findSegmentIndexByTime(normalizedTranscript, range.end));
            const rangeMatch = findTextInTranscript(normalizedTranscript, quoteText, transcriptIndex, {
              startIdx,
              strategy: 'all',
              minSimilarity: 0.75,
              maxSegmentWindow: Math.min(20, endIdx - startIdx + 5),
            });
            if (rangeMatch && rangeMatch.startSegmentIdx <= endIdx + 2) {
              const segment = segmentFromMatch(rangeMatch, normalizedTranscript, quoteText);
              if (segment) hydratedSegments = [segment];
            }
          }
        }
      }

      if (!hydratedSegments.length) {
        const segment = segmentFromTimestamp(topic.quote?.timestamp, normalizedTranscript, topic.quote?.text);
        if (segment) hydratedSegments = [segment];
      }

      if (!hydratedSegments.length) {
        const segment = fallbackSegment(normalizedTranscript);
        if (segment) hydratedSegments = [segment];
      }
    }

    return {
      ...topic,
      segments: hydratedSegments,
      duration: hydratedSegments.length ? computeDuration(hydratedSegments) : topic.duration,
    };
  });
}

export function extractThemes(candidates: TopicCandidate[] = []): string[] {
  if (!candidates.length) return [];
  const counts = new Map<string, number>();
  for (const c of candidates) {
    const words = c.title
      .toLowerCase()
      .split(/[^a-z0-9\u4e00-\u9fff]+/)
      .filter((w) => w.length > 3);
    for (const w of words) counts.set(w, (counts.get(w) ?? 0) + 1);
  }
  return [...counts.entries()]
    .filter(([, n]) => n >= 2)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 5)
    .map(([w]) => w);
}

export function findCandidatesForTheme(
  theme: string,
  candidates: TopicCandidate[],
): TopicCandidate[] {
  const t = theme.toLowerCase();
  return candidates.filter((c) => c.title.toLowerCase().includes(t));
}
